nucypher/datastore/base.py

```diff
@@ -0,0 +1,212 @@
+        if field_value is None:
+            raise AttributeError(f"No {record_field} record found for ID: {self._record_id}.")
+        return field_value
+
+    def __write_raw_record(self, record_field: str, value: bytes) -> None:
+        """
+        Writes a raw record, as bytes, to the database given a `record_field`
+        and a `value`.
+        If the record is unable to be written, this method raises a `DBWriteError`.
+        """
+        key = self.__storagekey.format(record_field=record_field, record_id=self._record_id).encode()
+        if not self.__db_tx.put(key, value, overwrite=True):
+            raise DBWriteError(f"Couldn't write the record (key: {key}) to the database.")
+
+    def __delete_record(self, record_field: str) -> None:
+        """
+        Deletes the record from the datastore.
+        """
+        key = self.__storagekey.format(record_field=record_field, record_id=self._record_id).encode()
+        if not self.__db_tx.delete(key) and self.__db_tx.get(key) is not None:
+            # We do this check to ensure that the key was actually deleted.
+            raise DBWriteError(f"Couldn't delete the record (key: {key}) from the database.")
+
+    def __get_record_field(self, attr: str) -> 'RecordField':
+        """
+        Uses `getattr` to return the `RecordField` object for a given
+        attribute.
+        These objects are accessed via class attrs as `_{attribute}`. If the
+        `RecordField` doesn't exist for a given `attr`, then this method will
+        raise a `TypeError`.
+        """
+        try:
+            record_field = getattr(self, f'_{attr}')
+        except AttributeError:
+            raise TypeError(f'No valid RecordField found on {self} for {attr}.')
+        return record_field
+
+    def __eq__(self, other):
+        """
+        WARNING: Records are only considered unique per their record IDs in this method.
+        In some cases (Workorder vs PolicyArrangement), these records may have
+        the same IDs. This comparison is useful when iterating over the 
+        datastore _as a subset of record type_ (as we do in queries), but not
+        useful when comparing two records of different type.
+        """
+        # We want to be able to compare this record to other IDs for the
+        # set operation in the query without instantiating a whole record.
+        if type(other) in (int, str):
+            return self._record_id == other
+        return self._record_id == other._record_id
+
+    def __hash__(self):
+        """
+        WARNING: Records are only considered unique per their record IDs in this method.
+        In some cases (Workorder vs PolicyArrangement), these records may have
+        the same IDs. This comparison is useful when iterating over the 
+        datastore _as a subset of record type_ (as we do in queries), but not
+        useful when comparing two records of different type.
+        """
+        return hash(self._record_id)
```

Comment:
I'm a little worried about this becoming the source of a bug.  why is `__eq__` not enough? ( @jMyles please advise )

Reply:
I... guess I agree that this is a very plausible source of a bug.  And the benefit is not immediately obvious to me.  Why not add one or more objects to the hash that make it decidedly unique?

Reply:
As we discussed in Discord, this is decidedly unique as long as record IDs are unique. I think this is a reasonable assumption when designing models and is incredibly obvious when not being done.